src/imagescache.cpp

```diff
@@ -0,0 +1,291 @@
+    return _images.value(id).cached;
+}
+
+//suppose to be run in non-gui thread
+QImage ImagesCache::image(const QString &id)
+{
+    QImage image_;
+    //qDebug() << "request for image" << id << _images.contains(id) << _images.value(id).cached;
+    if (_images.contains(id)) {
+        if (_images.value(id).cached) {
+            image_.load(_cache + QDir::separator() + _images.value(id).fileName);
+        } else {
+            emit requestImageViaHttp(id);
+        }
+    } else {
+        image_.load(QStringLiteral("://icons/smile.gif"));
+    }
+    return image_;
+}
+
+void ImagesCache::onImageRequestedViaHttp(const QString &id)
+{
+    QNetworkRequest req_ = QNetworkRequest(QUrl(_images.value(id).url));
+    req_.setAttribute(QNetworkRequest::User, QVariant(id));
+    QNetworkReply* reply = _qnam.get(req_);
+    connect(reply, &QNetworkReply::finished, this, &ImagesCache::onImageRequestFinished);
+}
+
+bool ImagesCache::parseSlackMojis(const QByteArray &data)
+{
+    auto doc = QGumboDocument::parse(data.data());
+    auto root = doc.rootNode();
+    auto nodes = root.getElementsByClassName(QStringLiteral("group"));
+    //    qDebug() << "groups" << nodes.size();
+
+    for (const auto& node: nodes) {
+        QString title_ = node.getElementsByClassName(QStringLiteral("title")).at(0).innerText().trimmed();
+        //qDebug() << "title" << title_;
+        auto emojis = node.getElementsByClassName(QStringLiteral("emojis")).front().children();
+        //qDebug() << "emojis" << emojis.size();
+        for (const auto& emoji: emojis) {
+            auto aclass = emoji.getElementsByTagName(HtmlTag::A).front();
+            ImageInfo info_;
+            info_.pack = title_;
+            info_.name = emoji.getAttribute(QStringLiteral("title")).trimmed();
+            QString url_ = aclass.getAttribute(QStringLiteral("href")).trimmed();
+            if (!url_.startsWith(QStringLiteral("http"))) {
+                url_.prepend("https://slackmojis.com");
+            }
+            info_.url = QUrl(url_);
+            info_.fileName = aclass.getAttribute(QStringLiteral("download")).trimmed();
+            info_.cached = false; //to check later
+            _images[info_.name] = info_;
+            //qDebug() << "adding emoji:" << info_.pack << info_.name << info_.url << info_.fileName;
+        }
+
+    }
+    return true;
+}
+
+bool ImagesCache::parseEmojiCheactSheet(const QByteArray &data)
+{
+    auto doc = QGumboDocument::parse(data.data());
+    auto root = doc.rootNode();
+    auto nodes = root.getElementsByTagName(HtmlTag::UL);
+    //qDebug() << "emojis" << nodes.size();
+
+    for (const auto& node: nodes) {
+        const QStringList & list_ = node.classList();
+        if (list_.isEmpty()) {
+            continue;
+        }
+        QString title_ = node.classList().at(0).trimmed();
+        //qDebug() << "title" << title_;
+        auto emojis = node.getElementsByClassName(QStringLiteral("emoji"));
```

Comment:
not a big fan of auto, can't see the type when reading the code.

the Qt coding guidelines has some exceptions for when it can be used (basically when it is obvious from the right hand side what the type is), but personally I prefer to avoid it always, and rather add «using» for various huge types.